### PR TITLE
Adds /login and authentication middleware

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: clients/crm-client.js [hunk 1]
```diff
@@ -173,8 +173,8 @@ class CRMClient {
    * Executes POST request against special batch endpoint ('/$batch'), with special batch body.
    * Returns the parsed response content on success, or false on failure.
    */
-  async doBatchPost(query, fetchXML) {
-    const batchBody = CRMClient.makeBatchBody(query, fetchXML);
+  async doBatchPost(query) {
+    const batchBody = CRMClient.makeBatchBody(query);
```
Comment (allthesignals) [inline clients/crm-client.js:177]: Are the changes here just cleanup for lint or something?

File: middleware/authenticate.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  try {
+    const sessionData = jwt.verify(token, CRM_SIGNING_SECRET);
+
+    // Unauth -- expired. Return un-auth response here?
+    if (sessionData.exp <= Date.now() / 1000) {
+      res.clearCookie('token');
+      next();
+    }
+
+    // Lookup the session from id. TODO: create a session store (redis?)
```
Comment (allthesignals) [inline middleware/authenticate.js]: No need for a session store, though, right? JWT stores data for us

File: middleware/authenticate.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  const { token } = req.cookies;
+
+  if (!token) {
+    res.clearCookie('token');
+    next();
+    return;
+  }
+
+  try {
+    const sessionData = jwt.verify(token, CRM_SIGNING_SECRET);
```
Comment (allthesignals) [inline middleware/authenticate.js:23]: The original cookie'd token here would have been signed by _our_ API (as opposed to the NYC.ID token)

File: routes/login.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    if (e instanceof BadRequestError) {
+      res.status(e.status).send({ errors: [{ code: e.code, detail: e.message }] });
+    }
+
+    console.log(e);
+    res.status(500).send({ errors: [{ detail: 'Unable to login' }] });
+  }
+});
+
+function validateNYCIDToken(token) {
```
Comment (allthesignals) [inline routes/login.js:46]: awesome!

**Review by allthesignals — APPROVED** [review]
for @allthesignals TODO add tests for auth

File: middleware/authenticate.js (hunk 3 above)
Comment (ghost) [reply middleware/authenticate.js:23]: yes, thats why we use the thing im calling CRM_SIGNING_SECRET to decode, not NYCID_CONSOLE_PASSWORD :)

File: clients/crm-client.js (hunk 1 above)
Comment (ghost) [reply clients/crm-client.js:177]: no i changed the function signature to make it match the `doGet` function b/c there was no real logic to keeping them separate. i think this is less error prone, and i end up just concat-ing them anyway.

File: middleware/authenticate.js (hunk 2 above)
Comment (ghost) [reply middleware/authenticate.js]: yes yes, i removed this, will cleanup the comments

**Pull request merged**